### PR TITLE
Move leaky_relu to Aten(CPU, CUDA)

### LINKED ISSUE #24583: Migrate `leaky_relu` and `leaky_relu_` from the TH to Aten (CUDA)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

### PR DESCRIPTION
**Pull request by XiaobingSuper**
@VitalyFedyunin, This PR is about port LeakyReLU activation to Aten:
**Test script:**
```
import torch
import torch.nn as nn
import time

torch.manual_seed(0)
def _time():
    if torch.cuda.is_available():
        torch.cuda.synchronize()
    return time.time()

device = "cpu"
m = nn.LeakyReLU()
if torch.cuda.is_available():
    device = "cuda"
    m = m.cuda()

#warm up
for n in [100, 10000]:
    input = torch.randn(128, n, requires_grad=True, device=device)
    grad_output = torch.ones(128, n, device=device)
    for i in range(1000):
        output = m(input)
        output.backward(grad_output)

for n in [100, 10000]:
    fwd_t = 0
    bwd_t = 0
    input = torch.randn(128, n, requires_grad=True, device=device)
    grad_output = torch.ones(128, n, device=device)
    for i in range(10000):
        t1 = _time()
        output = m(input)
        t2 = _time()
        output.backward(grad_output)
        t3 = _time()
        fwd_t = fwd_t + (t2 -t1)
        bwd_t = bwd_t + (t3 - t2)
    fwd_avg = fwd_t / 10000 * 1000
    bwd_avg = bwd_t / 10000 * 1000
    print("input size(128, %d) forward time is %.2f (ms); backwad avg time is %.2f (ms)."
          % (n, fwd_avg, bwd_avg))
```
Test Device: CPU: skx-8180, GPU: Tesla P40.
Perfromance:
Before:
```
GPU:
input size(128, 100) forward time is 0.05 (ms); backwad avg time is 0.11 (ms).
input size(128, 10000) forward time is 0.06 (ms); backwad avg time is 0.17 (ms).

CPU:
OMP_NUM_THREADS=56
input size(128, 100) forward time is 0.05 (ms); backwad avg time is 0.14 (ms).
input size(128, 10000) forward time is 4.21 (ms); backwad avg time is 8.02 (ms).
OMP_NUM_THREADS=1
input size(128, 100) forward time is 0.02 (ms); backwad avg time is 0.07 (ms).
input size(128, 10000) forward time is 1.98 (ms); backwad avg time is 6.21 (ms)
```
After:
```
GPU:
input size(128, 100) forward time is 0.05 (ms); backwad avg time is 0.11 (ms).
input size(128, 10000) forward time is 0.06 (ms); backwad avg time is 0.17 (ms).

CPU:
OMP_NUM_THREADS=56
input size(128, 100) forward time is 0.02 (ms); backwad avg time is 0.04 (ms).
input size(128, 10000) forward time is 0.03 (ms); backwad avg time is 0.09 (ms).
OMP_NUM_THREADS=1
input size(128, 100) forward time is 0.01 (ms); backwad avg time is 0.02 (ms).
input size(128, 10000) forward time is 0.47 (ms); backwad avg time is 1.02 (ms).
```
How to set the numbers of thread? using following script:
```
num_threads=$1
script=$2
last_core=`expr $num_threads - 1`
echo "using $num_threads OMP threads"
echo "bind cores to 0~$last_core"
export OMP_NUM_THREADS=$num_threads
export KMP_AFFINITY=granularity=fine,compact,1,0
numactl --physcpubind=0-$last_core --membind=0 python $script
```
and run .**/run.sh num_threads test.py**.

Fixes #24583 #24584 #24720 #24721

### CODE REVIEW

**Comment by XiaobingSuper** [comment]
#24583 Migrate leaky_relu and leaky_relu_ from the TH to Aten (CUDA)
#24584 Migrate leaky_relu_backward from the TH to Aten (CUDA)
#24720 Migrate leaky_relu and leaky_relu_ from the TH to Aten (CPU)
#24721 Migrate leaky_relu_backward from the TH to Aten (CPU)

**Comment by XiaobingSuper** [comment]
@xuhdev, can you help review it?

File: aten/src/ATen/native/cpu/Activation.cpp [hunk 1]
```diff
@@ -227,6 +227,45 @@ void hardshrink_backward_cpu_kernel(TensorIterator& iter, Scalar lambd) {
...
+    scalar_t negval = negval_.to<scalar_t>();
+    Vec negval_v = Vec(negval);
+    cpu_kernel_vec(
+        iter,
+        [&](scalar_t a) -> scalar_t {
+          return a > scalar_t(0) ? a : a * negval;
+        },
+        [&](Vec a) -> Vec {
+          auto r = Vec::blendv(negval_v, one_vec, a > zero_vec);
+          return a * r;
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/Activation.cpp]: All PR looks good overall, but this vectorization might be subject to precision loss. I need some time to double check, or you can introduce tests that will ensure me.

Comment (XiaobingSuper) [reply aten/src/ATen/native/cpu/Activation.cpp]: @VitalyFedyunin, thanks, can you tell me which one may cause the precision loss? **blendv** or “*”.

Comment (xuhdev) [reply aten/src/ATen/native/cpu/Activation.cpp]: I guess @VitalyFedyunin meant `*` because of the worry of `x * 1.0 != x`?

Comment (VitalyFedyunin) [reply aten/src/ATen/native/cpu/Activation.cpp]: Aha, everything looks fine.

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by VitalyFedyunin** [comment]
Requires rebase

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, just rebase the code.

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit c4fd97f:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=c4fd97f9d5b29a176034137e35b1cf2dc47f3cc7).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/3913180) pytorch_windows_build (1/1)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=72730049))
```
        ]
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/device_vector.h(55): note: see reference to class template instantiation 'thrust::detail::vector_base<T,Alloc>' being compiled
        with
        [
            T=double,
            Alloc=thrust::device_allocator<double>
        ]
C:/Users/circleci/project/aten/src\THC/generic/THCTensorMode.cu(30): note: see reference to class template instantiation 'thrust::device_vector<double,thrust::device_allocator<double>>' being compiled
-- Removing C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/./torch_cuda_generated_THCTensorMode.cu.obj
"C:/Program Files/CMake/bin/cmake.exe" -E remove C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/./torch_cuda_generated_THCTensorMode.cu.obj
CMake Error at torch_cuda_generated_THCTensorMode.cu.obj.Release.cmake:281 (message):
  Error generating file
  C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/./torch_cuda_generated_THCTensorMode.cu.obj


[1999/2729] cmd.exe /C "cd /D C:\Users\circleci\project\build\caffe2\CMakeFiles\torch_cuda.dir\__\aten\src\THC\generated && "C:\Program Files\CMake\bin\cmake.exe" -E make_directory C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/. && "C:\Program Files\CMake\bin\cmake.exe" -D verbose:BOOL=ON -D build_configuration:STRING=Release -D generated_file:STRING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMathReduceInt.cu.obj -D generated_cubin_file:STRING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMathReduceInt.cu.obj.cubin.txt -P C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/torch_cuda_generated_THCTensorMathReduceInt.cu.obj.Release.cmake"
-- Removing C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMathReduceInt.cu.obj
"C:/Program Files/CMake/bin/cmake.exe" -E remove C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMathReduceInt.cu.obj
-- Generating dependency file: C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/torch_cuda_generated_THCTensorMathReduceInt.cu.obj.NVCC-depend
C:/Users/circleci/project/build/win_tmp/bin/nvcc -M -D__CUDACC__ C:/Users/circleci/project/aten/src/THC/generated/THCTensorMathReduceInt.cu -o C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/torch_cuda_generated_THCTensorMathReduceInt.cu.obj.NVCC-depend -ccbin cl.exe -m64 -Dtorch_cuda_EXPORTS -DTORCH_CUDA_BUILD_MAIN_LIB -DTH_BLAS_MKL -D_OPENMP_NOFORCE_MANIFEST -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -D_CRT_SECURE_NO_DEPRECATE=1 -DWIN32_LEAN_AND_MEAN -DMAGMA_V2 -DAT_PARALLEL_OPENMP=1 -Xcompiler ,\"/DWIN32\",\"/D_WINDOWS\",\"/GR\",\"/w\",\"/MP\",\"/bigobj\",\"-openmp:experimental\",\"-DHAVE_AVX_CPU_DEFINITION\",\"-DHAVE_AVX2_CPU_DEFINITION\",\"/MD\",\"/O2\",\"/Ob2\",\"/w\",\"/MP\",\"/bigobj\" -Xcompiler /w -w -DONNX_NAMESPACE=onnx_torch -gencode arch=compute_75,code=sm_75 -Xcudafe --diag_suppress=cc_clobber_ignored -Xcudafe --diag_suppress=integer_sign_change -Xcudafe --diag_suppress=useless_using_declaration -Xcudafe --diag_suppress=set_but_not_used -Xcompiler -MD --expt-relaxed-constexpr --expt-extended-lambda -Xcompiler /wd4819 -Xcompiler /wd4503 -Xcompiler /wd4190 -Xcompiler /wd4244 -Xcompiler /wd4251 -Xcompiler /wd4275 -Xcompiler /wd4522 -Wno-deprecated-gpu-targets --expt-extended-lambda -gencode arch=compute_75,code=sm_75 -DCUDA_HAS_FP16=1 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ -DNVCC "-IC:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include" -IC:/Users/circleci/project/build/aten/src -IC:/Users/circleci/project/aten/src -IC:/Users/circleci/project/build -IC:/Users/circleci/project -IC:/Users/circleci/project/cmake/../third_party/googletest/googlemock/include -IC:/Users/circleci/project/cmake/../third_party/googletest/googletest/include -IC:/Users/circleci/project/third_party/protobuf/src -IC:/Users/circleci/project/build/win_tmp/mkl/include -IC:/Users/circleci/project/cmake/../third_party/benchmark/include -IC:/Users/circleci/project/cmake/../third_party/eigen -IC:/Jenkins/Miniconda3/include -IC:/Jenkins/Miniconda3/lib/site-packages/numpy/core/include -IC:/Users/circleci/project/cmake/../third_party/pybind11/include -I/opt/rocm/hip/include -I/include -IC:/Users/circleci/project/cmake/../third_party/cub -IC:/Users/circleci/project/build/caffe2/contrib/aten -IC:/Users/circleci/project/third_party/onnx -IC:/Users/circleci/project/build/third_party/onnx -IC:/Users/circleci/project/third_party/foxi -IC:/Users/circleci/project/build/third_party/foxi -IC:/Users/circleci/project/build/win_tmp/magma/include "-IC:/Program Files/NVIDIA Corporation/NvToolsExt/include" -IC:/Users/circleci/project/build/caffe2/aten/src/TH -IC:/Users/circleci/project/aten/src/TH -IC:/Users/circleci/project/build/caffe2/aten/src/THC -IC:/Users/circleci/project/aten/src/THC -IC:/Users/circleci/project/aten/src/THCUNN -IC:/Users/circleci/project/aten/src/ATen/cuda -IC:/Users/circleci/project/build/caffe2/aten/src -IC:/Users/circleci/project/aten/../third_party/catch/single_include -IC:/Users/circleci/project/aten/src/ATen/.. -IC:/Users/circleci/project/build/caffe2/aten/src/ATen -IC:/Users/circleci/project/c10/cuda/../.. -IC:/Users/circleci/project/c10/../ -IC:/Users/circleci/project/caffe2/../torch/csrc/api -IC:/Users/circleci/project/caffe2/../torch/csrc/api/include
THCTensorMathReduceInt.cu
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin

**Review by VitalyFedyunin — APPROVED** [review]
